runner/src/runner.rs

```diff
@@ -0,0 +1,631 @@
+    fs::create_dir_all(&runner.download_path())?;
+    let release = runner.try_get_release(false).await?.expect("No current release");
+    // WARNING: This will overwrite any pre-existing binary with the same name
+    // TODO: Check if a release with the same version is already at the `download_path`
+    runner.download_binary(release).await?;
+
+    runner.run_binary().await?;
+
+    loop {
+        if let Some(new_release) = runner.try_get_release(false).await? {
+            let maybe_downloaded_release = runner.downloaded_release();
+            let downloaded_release = maybe_downloaded_release.as_ref().ok_or(Error::NoDownloadedRelease)?;
+            if new_release.uri != downloaded_release.release.uri {
+                // Wait for child process to finish completely.
+                // To ensure there can't be two vault processes using the same Bitcoin wallet.
+                runner.terminate_proc_and_wait()?;
+
+                // Delete old release
+                runner.delete_downloaded_release()?;
+
+                // Download new release
+                runner.download_binary(new_release).await?;
+
+                // Run the downloaded release
+                runner.run_binary().await?;
+            }
+        }
+        tokio::time::sleep(BLOCK_TIME).await;
+    }
+}
+
+#[async_trait]
+pub trait RunnerExt {
+    fn parachain_rpc(&self) -> &WebsocketClient;
+    fn vault_args(&self) -> &Vec<String>;
+    fn child_proc(&mut self) -> &mut Option<Child>;
+    fn set_child_proc(&mut self, child_proc: Child);
+    fn downloaded_release(&self) -> &Option<DownloadedRelease>;
+    fn set_downloaded_release(&mut self, downloaded_release: DownloadedRelease);
+    fn download_path(&self) -> &PathBuf;
+    fn set_download_path(&mut self, download_path: PathBuf);
+    async fn try_get_release(&self, pending: bool) -> Result<Option<ClientRelease>, Error>;
+    async fn download_binary(&mut self, release: ClientRelease) -> Result<(), Error>;
+    fn uri_to_bin_path(&self, uri: &str) -> Result<(String, PathBuf), Error>;
+    fn delete_downloaded_release(&mut self) -> Result<(), Error>;
+    async fn run_binary(&mut self) -> Result<(), Error>;
+    fn terminate_proc_and_wait(&mut self) -> Result<(), Error>;
+    async fn get_request_bytes(&self, url: String) -> Result<Bytes, Error>;
+}
+
+#[async_trait]
+impl RunnerExt for Runner {
+    fn parachain_rpc(&self) -> &WebsocketClient {
+        &self.parachain_rpc
+    }
+
+    fn vault_args(&self) -> &Vec<String> {
+        &self.vault_args
+    }
+
+    fn child_proc(&mut self) -> &mut Option<Child> {
+        &mut self.child_proc
+    }
+
+    fn set_child_proc(&mut self, child_proc: Child) {
+        self.child_proc = Some(child_proc);
+    }
+
+    fn downloaded_release(&self) -> &Option<DownloadedRelease> {
+        &self.downloaded_release
+    }
+
+    fn set_downloaded_release(&mut self, downloaded_release: DownloadedRelease) {
```

Comment:
Nitpick: imo it's not really necessary to have these kind of functions as trait item: you don't mock them anywhere and it's only called from within the Runner itself. Imo you can just inline this where it's called. But this is mostly a matter of preference I guess, so feel free to ignore

Reply:
These getters and setters are being mocked though: https://github.com/savudani8/interbtc-clients/blob/5a355afbac84074e63a86bb5a72e9f6511337109/runner/src/runner.rs#L615

Reply:
I can't see where `set_downloaded_release` is mocked @savudani8...